### PR TITLE
Save Validator Public Key -> Index in DB

### PR DESCRIPTION
**Pull request by terenc3t**
Currently to retrieve a validator index it's an O(n) operation to loop through entire validator registry. Given this is a frequent serving operation of the RPC server (i.e. anyone can query its validator index from inputting public key) It's best to use something O(1)

Key changes:
- Implemented DB wrappers for _save_, _get_ and _delete_ validator pub key -> index pair
- Updated chain start routine to save newly activated validator's pub key -> index in DB
- Updated RPC server functions
- Verified e2e works

### CODE REVIEW

**Comment by rauljordan** [comment]
@terenc3t it seems to me more natural to store the validator index in the validator client's persistent storage, which can then be persisted across sessions and used across all RPC requests the validator makes to the beacon node. In this PR it seems you're storing it in the beacon node. I'd rather give this storage overhead to the validator itself. I think this is also the direction the validator spec is leaning towards.

**Comment by terenc3t** [comment]
> @terenc3t it seems to me more natural to store the validator index in the validator client's persistent storage, which can then be persisted across sessions and used across all RPC requests the validator makes to the beacon node. In this PR it seems you're storing it in the beacon node. I'd rather give this storage overhead to the validator itself. I think this is also the direction the validator spec is leaning towards.

This is storing all validators pub keys to their respective indexes, there's another PR coming that's to store validators own index in the validator client's persistent storage.

Why do we need this in the beacon node?? 

1.) Any validator with a pub key can query its index with O(1) look up versus O(n) (e.g. `ValidatorIdx` in validator package)
2.) Currently a validator retrieves assignment using public key, that's an additional O(n) over head to convert to index, this alleviates that to O(1) over head

Overall I think there's enough benefit for a beacon node to have this in DB

File: beacon-chain/db/state.go [hunk 1]
```diff
@@ -48,6 +50,13 @@ func (db *BeaconDB) InitializeState(genesisTime uint64, deposits []*pb.Deposit)
 			return err
 		}
 
+		for i, validator := range beaconState.ValidatorRegistry {
+			h := hashutil.Hash(validator.Pubkey)
+			if err := validatorBkt.Put(h[:], []byte(strconv.Itoa(i))); err != nil {
```
Comment (prestonvanloon) [inline beacon-chain/db/state.go]: why do you have to convert the int to a string to a byte array? 

Maybe use the `encoding/binary` package? 
https://godoc.org/encoding/binary#PutUvarint

Comment (prestonvanloon) [reply beacon-chain/db/state.go]: By the way, I had to prove this to myself in a bench.

Results:
```
goos: linux
goarch: amd64
BenchmarkStrconv-4              30000000                47.3 ns/op
BenchmarkBinaryEncoding-4       300000000                5.77 ns/op
PASS
ok      _/tmp/bench     3.770s
```

Test:
```go
import (
  "encoding/binary"
  "strconv"
  "testing"
)

func BenchmarkStrconv(b *testing.B) {
  for i := 0; i < b.N; i++ {
    _ = []byte(strconv.Itoa(i))
  }
}

func BenchmarkBinaryEncoding(b *testing.B) {
  for i := uint64(0); i < uint64(b.N); i++ {
    buf := make([]byte, binary.MaxVarintLen64)
    _ = binary.PutUvarint(buf, i)
  }
}
```

Please fix here and everywhere else in the package. :)

Comment (terenc3t) [reply beacon-chain/db/state.go]: Excellent point!

**Comment by terenc3t** [comment]
Addressed your feedback @prestonvanloon

File: beacon-chain/db/validator.go [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+	"encoding/binary"
+	"fmt"
+
+	"github.com/boltdb/bolt"
+	"github.com/prysmaticlabs/prysm/shared/hashutil"
+)
+
+// SaveValidatorIndex accepts a public key and validator index and writes them to disk.
+func (db *BeaconDB) SaveValidatorIndex(pubKey []byte, index int) error {
+	h := hashutil.Hash(pubKey)
```
Comment (prestonvanloon) [inline beacon-chain/db/validator.go]: Why do we hash the pubkey?

File: beacon-chain/rpc/validator_server_test.go [hunk 3]
```diff
@@ -187,13 +181,20 @@ func TestNextEpochCommitteeAssignment_OK(t *testing.T) {
 	if err := db.UpdateChainHead(genesis, state); err != nil {
 		t.Fatalf("Could not save genesis state: %v", err)
 	}
+
+	var pubKey [96]byte
+	for i := 0; i < int(params.BeaconConfig().DepositsForChainStart); i++ {
+		copy(pubKey[:], strconv.Itoa(i))
```
Comment (prestonvanloon) [inline beacon-chain/rpc/validator_server_test.go]: `strconv` here again

File: beacon-chain/db/validator.go (hunk 2 above)
Comment (terenc3t) [reply beacon-chain/db/validator.go]: Yeah it's to save space, 96 bytes vs 32 bytes. That makes sense ya?

**Comment by terenc3t** [comment]
> LGTM, I would like a comment on why we are hashing the pubkey for the database key.
> 
> Is it to save disk space? What is the trade off and why?

It's to save disk space. 96 bytes (later 48 bytes) vs. 32 bytes

File: beacon-chain/db/validator.go (hunk 2 above)
Comment (prestonvanloon) [reply beacon-chain/db/validator.go]: Is it worth the performance cost though?

Comment (terenc3t) [reply beacon-chain/db/validator.go]: It's about `1206 ns/op` over head with the additional hashing, but the trade off is key size goes down to 32 bytes as opposed 96 bytes

```go
func BenchmarkWithHash(b *testing.B) {
	for i := uint64(0); i < uint64(b.N); i++ {
		buf := make([]byte, binary.MaxVarintLen64)
		n := binary.PutUvarint(buf, i)
		_ = hashutil.Hash(buf[n:])
	}
}
```
`BenchmarkWithHash-4   	 1000000	      1206 ns/op`

Comment (prestonvanloon) [reply beacon-chain/db/validator.go]: Thanks, I’m satisified with this answer

**Comment by prestonvanloon** [comment]
LGTM

**Pull request merged**